internal/resource/config/idp/adapter/idp_adapter_common.go

```diff
@@ -0,0 +1,188 @@
+	"github.com/pingidentity/terraform-provider-pingfederate/internal/resource/common/issuancecriteria"
+	"github.com/pingidentity/terraform-provider-pingfederate/internal/resource/common/pluginconfiguration"
+	"github.com/pingidentity/terraform-provider-pingfederate/internal/resource/common/resourcelink"
+	internaltypes "github.com/pingidentity/terraform-provider-pingfederate/internal/types"
+)
+
+// Define attribute types for object types
+var (
+	// May move some of this into common package if future resources need this
+	attributesAttrType = map[string]attr.Type{
+		"name":      types.StringType,
+		"pseudonym": types.BoolType,
+		"masked":    types.BoolType,
+	}
+
+	attributeContractAttrTypes = map[string]attr.Type{
+		"core_attributes": types.SetType{
+			ElemType: types.ObjectType{
+				AttrTypes: attributesAttrType,
+			},
+		},
+		"core_attributes_all": types.SetType{
+			ElemType: types.ObjectType{
+				AttrTypes: attributesAttrType,
+			},
+		},
+		"extended_attributes": types.SetType{
+			ElemType: types.ObjectType{
+				AttrTypes: attributesAttrType,
+			},
+		},
+		"unique_user_key_attribute": types.StringType,
+		"mask_ognl_values":          types.BoolType,
+		"inherited":                 types.BoolType,
+	}
+
+	attributeContractDataSourceAttrTypes = map[string]attr.Type{
+		"core_attributes": types.SetType{
+			ElemType: types.ObjectType{
+				AttrTypes: attributesAttrType,
+			},
+		},
+		"extended_attributes": types.SetType{
+			ElemType: types.ObjectType{
+				AttrTypes: attributesAttrType,
+			},
+		},
+		"unique_user_key_attribute": types.StringType,
+		"mask_ognl_values":          types.BoolType,
+		"inherited":                 types.BoolType,
+	}
+
+	attributeMappingAttrTypes = map[string]attr.Type{
+		"attribute_sources": types.ListType{
+			ElemType: types.ObjectType{
+				AttrTypes: attributesources.ElemAttrType(),
+			},
+		},
+		"attribute_contract_fulfillment": attributecontractfulfillment.MapType(),
+		"issuance_criteria": types.ObjectType{
+			AttrTypes: issuancecriteria.AttrType(),
+		},
+		"inherited": types.BoolType,
+	}
+
+	extendedAttributesDefault, _ = types.SetValue(types.ObjectType{
+		AttrTypes: attributesAttrType,
+	}, nil)
+)
+
+type idpAdapterModel struct {
+	AuthnCtxClassRef    types.String `tfsdk:"authn_ctx_class_ref"`
+	Id                  types.String `tfsdk:"id"`
+	AdapterId           types.String `tfsdk:"adapter_id"`
+	Name                types.String `tfsdk:"name"`
+	PluginDescriptorRef types.Object `tfsdk:"plugin_descriptor_ref"`
+	ParentRef           types.Object `tfsdk:"parent_ref"`
+	Configuration       types.Object `tfsdk:"configuration"`
+	AttributeMapping    types.Object `tfsdk:"attribute_mapping"`
+	AttributeContract   types.Object `tfsdk:"attribute_contract"`
+}
+
+func readIdpAdapterResponse(ctx context.Context, r *client.IdpAdapter, state *idpAdapterModel, plan idpAdapterModel, isResource bool) diag.Diagnostics {
```

Comment:
You could make plan a pointer, so that we can pass it in as nil for the data source

Reply:
in latest commit